File: src/test/cc/wfa/virtual_people/common/field_filter/has_filter_test.cc
// Copyright 2021 The Cross-Media Measurement Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "gmock/gmock.h"
#include "google/protobuf/text_format.h"
#include "gtest/gtest.h"
#include "src/main/proto/wfa/virtual_people/common/field_filter.pb.h"
#include "src/test/cc/testutil/matchers.h"
#include "src/test/cc/testutil/status_macros.h"
#include "src/test/cc/wfa/virtual_people/common/field_filter/test/test.pb.h"
#include "wfa/virtual_people/common/field_filter/field_filter.h"

namespace wfa_virtual_people {
namespace {

using ::wfa::StatusIs;
using ::wfa_virtual_people::test::TestProto;

// This function is required to test the FieldFilter still works when the
// FieldFilterProto is out of scope.
absl::StatusOr<std::unique_ptr<FieldFilter>> FilterFromProtoText(
    absl::string_view proto_text) {
  FieldFilterProto field_filter_proto;
  EXPECT_TRUE(google::protobuf::TextFormat::ParseFromString(
      std::string(proto_text), &field_filter_proto));
  return FieldFilter::New(TestProto().GetDescriptor(), field_filter_proto);
}

TEST(HasFilterTest, TestNoName) {
  FieldFilterProto field_filter_proto;
  ASSERT_TRUE(
      google::protobuf::TextFormat::ParseFromString(R"pb(
                                                      op: HAS
                                                    )pb",
                                                    &field_filter_proto));
  EXPECT_THAT(FieldFilter::New(TestProto().GetDescriptor(), field_filter_proto)
                  .status(),
              StatusIs(absl::StatusCode::kInvalidArgument, ""));
}

TEST(HasFilterTest, TestInvalidName) {
  FieldFilterProto field_filter_proto;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      R"pb(
        name: "a.b.invalid_field" op: HAS
      )pb",
      &field_filter_proto));
  EXPECT_THAT(FieldFilter::New(TestProto().GetDescriptor(), field_filter_proto)
                  .status(),
              StatusIs(absl::StatusCode::kInvalidArgument, ""));
}

TEST(HasFilterTest, TestDisallowedRepeatedInThePath) {
  FieldFilterProto field_filter_proto;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      R"pb(
        name: "repeated_proto_a.b.int32_value" op: HAS
      )pb",
      &field_filter_proto));
  EXPECT_THAT(FieldFilter::New(TestProto().GetDescriptor(), field_filter_proto)
                  .status(),
              StatusIs(absl::StatusCode::kInvalidArgument, ""));
}

TEST(HasFilterTest, TestNonRepeated) {
  ASSERT_OK_AND_ASSIGN(std::unique_ptr<FieldFilter> field_filter,
                       FilterFromProtoText(R"pb(
                         name: "a.b.int32_value" op: HAS
                       )pb"));

  TestProto test_proto_1;
  ASSERT_TRUE(
      google::protobuf::TextFormat::ParseFromString(R"pb(
                                                      a { b { int32_value: 1 } }
                                                    )pb",
                                                    &test_proto_1));
  EXPECT_TRUE(field_filter->IsMatch(test_proto_1));

  TestProto test_proto_2;
  ASSERT_TRUE(
      google::protobuf::TextFormat::ParseFromString(R"pb(
                                                      a { b { int64_value: 1 } }
                                                    )pb",
                                                    &test_proto_2));
  EXPECT_FALSE(field_filter->IsMatch(test_proto_2));

  TestProto test_proto_3;
  EXPECT_FALSE(field_filter->IsMatch(test_proto_3));
}

TEST(HasFilterTest, TestRepeated) {
  ASSERT_OK_AND_ASSIGN(std::unique_ptr<FieldFilter> field_filter,
                       FilterFromProtoText(R"pb(
                         name: "a.b.int32_values" op: HAS
                       )pb"));

  TestProto test_proto_1;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      R"pb(
        a { b { int32_values: 1 } }
      )pb",
      &test_proto_1));
  EXPECT_TRUE(field_filter->IsMatch(test_proto_1));

  TestProto test_proto_2;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      R"pb(
        a { b { int32_values: 1 int32_values: 2 } }
      )pb",
      &test_proto_2));
  EXPECT_TRUE(field_filter->IsMatch(test_proto_2));

  TestProto test_proto_3;
  ASSERT_TRUE(
      google::protobuf::TextFormat::ParseFromString(R"pb(
                                                      a { b { int32_value: 1 } }
                                                    )pb",
                                                    &test_proto_3));
  EXPECT_FALSE(field_filter->IsMatch(test_proto_3));

  TestProto test_proto_4;
  EXPECT_FALSE(field_filter->IsMatch(test_proto_4));
}

}  // namespace
}  // namespace wfa_virtual_people

File: src/test/kotlin/org/wfanet/virtualpeople/common/fieldfilter/HasFilterTest.kt
// Copyright 2022 The Cross-Media Measurement Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.wfanet.virtualpeople.common.fieldfilter

import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertTrue
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import org.wfanet.virtualpeople.common.FieldFilterProto.Op
import org.wfanet.virtualpeople.common.fieldFilterProto
import org.wfanet.virtualpeople.common.test.TestProto
import org.wfanet.virtualpeople.common.test.testProto
import org.wfanet.virtualpeople.common.test.testProtoA
import org.wfanet.virtualpeople.common.test.testProtoB

@RunWith(JUnit4::class)
class HasFilterTest {

  @Test
  fun `filter without name should fail`() {
    val fieldFilter = fieldFilterProto { op = Op.HAS }
    val exception =
      assertFailsWith<IllegalStateException> {
        FieldFilter.create(TestProto.getDescriptor(), fieldFilter)
      }
    assertTrue(exception.message!!.contains("Name must be set"))
  }

  @Test
  fun `filter with invalid name should fail`() {
    val fieldFilter = fieldFilterProto {
      name = "a.b.invalid_field"
      op = Op.HAS
    }
    val exception =
      assertFailsWith<IllegalStateException> {
        FieldFilter.create(TestProto.getDescriptor(), fieldFilter)
      }
    assertTrue(exception.message!!.contains("The field name is invalid"))
  }

  @Test
  fun `filter with repeated field in the path should fail`() {
    val fieldFilter = fieldFilterProto {
      name = "repeated_proto_a.b.int32_value"
      op = Op.HAS
    }
    val exception =
      assertFailsWith<IllegalStateException> {
        FieldFilter.create(TestProto.getDescriptor(), fieldFilter)
      }
    assertTrue(exception.message!!.contains("Repeated field is not allowed in the path"))
  }

  @Test
  fun `has nonrepeated field`() {
    val fieldFilter = fieldFilterProto {
      name = "a.b.int32_value"
      op = Op.HAS
    }
    val filter = FieldFilter.create(TestProto.getDescriptor(), fieldFilter)
    val testProto1 = testProto { a = testProtoA { b = testProtoB { int32Value = 1 } } }
    assertTrue(filter.matches(testProto1))
    assertTrue(filter.matches(testProto1.toBuilder()))

    val testProto2 = testProto { a = testProtoA { b = testProtoB { int64Value = 1 } } }
    assertFalse(filter.matches(testProto2))
    assertFalse(filter.matches(testProto2.toBuilder()))

    val testProto3 = testProto {}
    assertFalse(filter.matches(testProto3))
    assertFalse(filter.matches(testProto3.toBuilder()))
  }

  @Test
  fun `has repeated field`() {
    val fieldFilter = fieldFilterProto {
      name = "a.b.int32_values"
      op = Op.HAS
    }
    val filter = FieldFilter.create(TestProto.getDescriptor(), fieldFilter)
    val testProto1 = testProto { a = testProtoA { b = testProtoB { int32Values.add(1) } } }
    assertTrue(filter.matches(testProto1))
    assertTrue(filter.matches(testProto1.toBuilder()))

    val testProto2 = testProto {
      a = testProtoA {
        b = testProtoB {
          int32Values.add(1)
          int32Values.add(2)
        }
      }
    }
    assertTrue(filter.matches(testProto2))
    assertTrue(filter.matches(testProto2.toBuilder()))

    val testProto3 = testProto { a = testProtoA { b = testProtoB { int32Value = 1 } } }
    assertFalse(filter.matches(testProto3))
    assertFalse(filter.matches(testProto3.toBuilder()))

    val testProto4 = testProto {}
    assertFalse(filter.matches(testProto4))
    assertFalse(filter.matches(testProto4.toBuilder()))
  }
}
